src/strategy.ts

```diff
@@ -57,38 +59,59 @@ export abstract class AbstractStrategy extends PassportStrategy {
       loggedOut: boolean;
     }) => {
       if (loggedOut) {
-        // Log out the current user no matter if we can verify the logged in user === logout requested user
-        req.logout();
-
-        // Check to see if we are logging out the user that is currently logged in to craft a proper IdP response
         if (profile != null) {
           if (this._saml == null) {
             throw new Error("Can't get logout response URL without a SAML provider defined.");
           }
 
-          const RelayState = req.query?.RelayState || req.body?.RelayState;
+          const getLogoutResponseUrl = this._saml.getLogoutResponseUrl;
+          // When logging out a user, use the consumer's `validate` function to check that
+          // the `profile` associated with the logout request resolves to the same user
+          // as the `profile` associated with the current session.
+          const verified = (err: Error | null, logoutUser?: User, info?: unknown) => {
+            if (err) {
+              return this.error(err);
+            }
+
+            if (!logoutUser) {
+              return this.fail(info, 401);
+            }
+
+            let userMatch = true;
+            try {
+              // Check to see if we are logging out the user that is currently logged in to craft a proper IdP response
+
+              // TODO: This may be a problem if there are any timestamps in here, should we do a `strictEqual` instead?
```

Comment:
You commented at https://github.com/node-saml/passport-saml/pull/619#discussion_r663131078 :
>...there is still a potential problem with the method I've chosen (see the TODO). I'll resolve this thread so we can start a new thread based on the new code.

I don't have enough knowledge about JS/TS stuff to discuss about this TODO

Reply:
It isn't so much about JS, it is about object comparison. Let's say that the `user` object returned by the consumer from a given `profile` object is something like this:

```js
{
  name: profile.nameID,
  lastLogIn: DateTime.now(),
}
```

A simple object like that makes the conversion from `profile` to `user` such that it would be different every time. That would break any chance of comparison and thus any chance of reporting a successful SLO.

The suggestion to do a `strictEqual` instead of a `strictDeepEqual` would only compare the top-level object properties, nothing nested. Thus, if there as a `datatime` property that is nested, that would not be used in the comparison. The assumption being that identity could be determined from the top-level properties. For example:

```js
{
  name: profile.nameID,
  loginData: {
    lastLogin: DateTime.now(),
    loginLocation: getLocationFromRequest(req)
  }
}
```

An object built like that would pass a `strictEqual`, but fail a `strictDeepEqual`.

Truthfully, it is almost looking like we'll need some other piece of information provided by the user for logout.

Reply:
There is now a second verify function so the user can do the calculation on finding the user to log out. If they return the same object that we have referenced internally, then we consider the logout successful, if not, then we consider the logout a failure.

Given there is a unique function for logout, the user can then cause whatever credential store, or whatever else they have, to do a logout as well and even handle their own sync issues.

Do you see any problems with this method @srd90 @markstos ?

Reply:
(Just to make it easier for those who might read this conversation later @cjbarth is referring to this modification [cade87c](https://github.com/node-saml/passport-saml/pull/619/commits/cade87c5a73f27561da036fd9101549f620b8518))

https://github.com/node-saml/passport-saml/blob/cade87c5a73f27561da036fd9101549f620b8518/src/strategy.ts#L86 check works if and only if express/passport was able to associate correct user to `req.user` in the first place. I.e. if there was proper session cookie along with the incoming `LogoutRequest` and thus `req.isAuthenticated` (or similar) is/would/could also return `true` and in this particular case this check would double check that authenticated session was created from login response impliciated by content of logoutrequest.

If session cookie was missing that check would fail always unless developer circumvents it by injecting object `foo` to `req.user` during this invocation https://github.com/node-saml/passport-saml/blob/cade87c5a73f27561da036fd9101549f620b8518/src/strategy.ts#L111 and then return same `foo` as a result of `VerifyWithRequest` callback invocation.

IMHO with [cade87c](https://github.com/node-saml/passport-saml/pull/619/commits/cade87c5a73f27561da036fd9101549f620b8518) approach `passport-saml` could return LogoutResponse with failure status without even consulting veifylogout callback if `req.user` was null or `req.isAuthenticated !== true`.

On the other hand current implementation would make it possible for above average developer to implement SAML compliant session management (i.e. session management which would support logging out without session cookie) because now it would be possible to have application specific code to be executed during LogoutRequest processing. Only problem would be that `passport-saml` uses `req.user` to determine whether it - `passport-saml` - should respond with Success to LogoutRequest. Workaround would  be to inject `foo` to `req.user` and return `foo` as a result of verifylogout callback. Not too clean API from above average developer point of view (and those who just don't care about correct SLO implementation would use same workaround to have logoutresponse with success status).

How about if verifylogout callback function would take same parameters as it takes [at the moment](https://github.com/node-saml/passport-saml/pull/619/commits/cade87c5a73f27561da036fd9101549f620b8518) but change its name to `executeSLOLogout` and if the implementation of that callback function returns a string `"I as a developer of this function solemnly swear and affirm that the session implicated by LogoutRequest is successfully terminated and it - session - is not reintroduced to session store even if there was concurrent request processing ongoing wich handled same user while LogoutRequest was processed"` (or maybe just `true`) passport-saml would respond with logoutresponse with `Success`. `passport-saml` could call behind the scenes req.logout() just before it sends logoutresponse but that call would be sort of "last effort". My point is that API could be nicer for those who are interested to implement proper IdP initiated SLO.

btw. now I remembered seeing this https://github.com/node-saml/passport-saml/pull/144 which is somewhat related to this pull request's discussion (possibility to execute custom logic after LogoutRequest signature verification but before response delivery). First answer in that pull request didn't address pull request author's second concern (idp initiated case) at all. I linked to that pull request just in case someone is interested browse related discussions.

Reply:
I interpreted your comments https://github.com/node-saml/passport-saml/pull/619#issuecomment-923005584 and https://github.com/node-saml/passport-saml/pull/619#discussion_r712246505 so that you planned to drop "deepStrictEqual" check (quote from latter linked comment):
>The more I think about it, the more I think that is the only way we can really let the user make the call. Then we'd just pass the user that we have and the profile that we got from the SLO request, and let the user of passport-saml return a go/no-go on the validity of the SLO request instead of trying to do that calculation blind in the passport-saml code.

Reply:
I see your point about a potentially fuzzy API here. My thinking was that we want to guarantee that the user that we're working with is the one we are telling the IdP that we're logging out. I don't want the consumer to be able to write a function like `() => true` and have SLO work. By forcing them to actually go get a `user` object (preferably from their user store, but possibly `() => req.user`, which should be a red flag to any programmer) and return it, and then we make sure it is 100% the same (so deep equal), we have strong assurances that we aren't lying to the IdP.

If there is no match, we still log out any current user, so that covers all the other cases, all without lying to the IdP.

I can see renaming the function, in fact, both of the functions aren't really a "verify"... but I'm not so sure they are "execute" either.

Reply:
@markstos , since you've approved this PR, I'm assuming your up to speed on this discussion and favor leaving these functions named "verify", so I'll land this PR. If there is a need to revisit, I'm more than open to making that simple refactor before we leave beta.